taichi/struct/struct_llvm.cpp

```diff
@@ -37,12 +37,17 @@ void StructCompilerLLVM::generate_types(SNode &snode) {
   std::vector<llvm::Type *> ch_types;
   for (int i = 0; i < snode.ch.size(); i++) {
     if (!snode.ch[i]->is_bit_level) {
-      // Bit-level SNodes do not really have a corresponding LLVM type
+//       Bit-level SNodes do not really have a corresponding LLVM type
       auto ch = get_llvm_node_type(module.get(), snode.ch[i].get());
       ch_types.push_back(ch);
+    } else {
+      DataType container_primitive_type(snode.physical_type);
+      ch_types.push_back(tlctx->get_data_type(container_primitive_type));
     }
   }
 
+  // if we don't push any thing into `ch_types` for bit_level SNode
+  // `ch_type` will be  a StructType with no element in it.
   auto ch_type =
       llvm::StructType::create(*ctx, ch_types, snode.node_type_name + "_ch");
```

Comment:
As I comment here, bit-level SNodes need to know their children types.

Reply:
In fact, they don't need to know their children types - and LLVM doesn't have corresponding mechanisms to generate these types. That's why we need bit pointers.